### ISSUE TITLE
Allow grading in parts with empty input.

### ISSUE THREAD

**Issue by shinuito** [FormulasQuestion/moodle-qtype_formulas#17]
Hi, 

This issue has come up at least twice that I can see on the Moodle forums, in [2018](https://moodle.org/mod/forum/discuss.php?d=369438) and recently in [September 2020](https://moodle.org/mod/forum/discuss.php?d=411103). Just to continue and move a discussion from there to here, which seems more appropriate. A brief summary of the issue below.

At the moment, it is very convenient to use the multiple input in parts of a formulas type question, both to align the display of the parts, and also to do combined grading on the parts of the question. For example I can align MCE type and input type boxes as below
<img width="324" alt="Screen Shot 2020-10-09 at 11 48 05" src="https://user-images.githubusercontent.com/26803867/95576425-579feb80-0a28-11eb-91c1-cb2a5a034424.png">
so that it looks like a student can input say an interval or set of real numbers with two elements as an answer. It can be a bit tricky to get students to input expressions like [a,b] into an input box and compare it, either with regular expressions or in another question type like STACK (even though they have commands like cc(a,b)  that they recognise). There's lots of tradeoffs between ease of input and ease of parsing the answer and grading it.

However there is a huge drawback. Suppose the four input boxes belong to the same part say {#1} and are displayed like {_0:choices:MCE}{_1} , {_2} {_3:choices2:MCE}, then since the answer is for example a numerical list [0,1,2,0] etc. If any of the four input boxes are left empty (as is likely to be the case, students do it all the time on paper assignments), the grading variables and marking are completely skipped and the answer is marked wrong. 

Now the possible work arounds are quite involved, especially if we assume the average person making the questions, just wants to use the WYSIWYG editors, learning minimal syntax for inserting parts/input boxes etc (The moodleformulas website is quite good at explaining these things). For instance one has to split the four input boxes above into four separate parts, manually edit the html display style, to get them to align, and now in theory you can get marks for each input even if other boxes are blank. That is unless the grading of the inputs are codependent. For instance imagine the answer is a set and you would like the student to be able to input either {0,1} or {1,0}. Now you can't really just split them into multiple parts, and some workarounds might involve javascript to copy inputs into hidden extra inputs.

Typically having to suggest one workaround to an issue will lead to many more workarounds needed. For instance, even getting the question to display the same using div inline, once the question is answered, the separate parts display their own feedback and that already destroys the html display, in a way it wouldn't if they were all one part

<img width="575" alt="Screen Shot 2020-10-09 at 11 48 23" src="https://user-images.githubusercontent.com/26803867/95576427-58388200-0a28-11eb-83a7-39a91c6a1692.png">

It would be great if there was a way to either ignore blank input boxes. One could have a special character/comparison or function that could check if the inputs were empty, but nonetheless proceed to the grading, so that parts with multiple inputs can be marked even if some inputs are empty.

If it is too much of a change that it could break old questions, then maybe allow this as a setting/toggle/checkbox which either keeps the old way, or allows empty inputs to still proceed to grading.

**Comment by PhilippImhof** [comment]
Thanks for reporting this. I am not quite sure what you wish to have. Currently, it is not possible for a correct answer to be empty or for an empty answer to be correct. This is due to the functioning of Moodle: an empty input field is the sign that the student has not entered a solution. If the true solution is indeed the empty string, it will be impossible to distinguish between students who knew the solution and chose to leave the field empty (i.e. "entering" the empty solution) and students who did not know what to put in the field. Is this what you were looking for?

When it comes to dealing with empty fields, there are two scenarios:

* When working with the behaviour "immediate feedback", one is normally in a learning situation rather than in an exam. In this case, a question that is considered incomplete will not be graded and the student will not get any feedback until they answer every part of the question. The idea is that an empty field will inevitably be wrong (see above), so in a learning situation it's useless to leave fields empty rather than to try something.
* In an exam situation, questions do not need to be complete in order to be graded, empty fields will simply be graded wrong.

In order to cover both scenarios, every question can decide what is needed in order for an answer to be "complete" and "gradable". In most cases, a question is considered complete if *all* parts are answered. And it is considered "gradable" if *at least one* field has been filled in. Both criteria can be changed freely by the question type's developers.

As it is IMHO generally a good thing to tell the students they have possibly missed a field, I would not want to change the default behaviour. So in order to get what you are (probably) looking for, one would need to add a checkbox in the question's edit form that allows this to be set on a per-question basis. I do not know whether that is a good idea, because it adds one more thing to an already complex form. (However, we can safely assume that most people working with that question type can probably handle complex stuff.)

**Comment by RedyAu** [comment]
We too had requests come in complaining about questions being unmarked when a student missed a field out of more than a dozen. In this case this had the opposite effect to the intended one: Rather than being a helpful reminder to fill in everything, they got no points even for the fields they _did_ fill in.

So a checkmark would be appreciated per question or part, which would allow for giving a mark of 0 for not filled parts, and still set the whole question as gradable.
This would achieve your quoted scenario:

> In an exam situation, questions do not need to be complete in order to be graded, empty fields will simply be graded wrong.

-- which, as of yet, is not the case (unless I got something wrong). The whole question is skipped for marking, not just the specific part.

**Comment by PhilippImhof** [comment]
@RedyAu What you are observing is probably related to #219 and, to a lesser extent, #213. There was indeed a bug that led to incomplete questions not being correctly graded in deferred feedback mode a.k.a. in an exam situation.

**Comment by RedyAu** [comment]
Great news! Sorry that I didn't check the issues more thoroughly...
I believe this solves the original premise of this issue, so it could be closed - of course I can't read the mind of @shinuito either.

**Comment by PhilippImhof** [comment]
No worries.

I don't think this solves what @shinuito had in mind. The original issue was opened for an older version of the plugin, quite a while before I introduced that bug…

**Comment by shinuito** [comment]
Yes, my original comment was from 2020, so I don't know how much the plugin has changed since then.

I think it was probably to do with how if you wanted to grade multiple inputs together in a single part, you couldn't award partial marks if a student left one of the inputs empty (as they sometimes do).

I think I was probably suggesting that one should still be able to allow for and grade empty input boxes.

- - - 

I ended up creating a lot of questions in STACK over the next four years, and I can say that allowing empty input boxes was useful in some kinds of questions. It has a placeholder value for empty answer boxes that allowed you to grade empty inputs even if it was part of grading multiple inputs together. Or in some cases a certain type of question it made sense to leave a certain input empty.

The example might be having a series, and asking if the series is divergent or convergent. If it's convergent which test would you use. If you used a p-series test, what value of p? You might want to leave the "p" box empty in some variants for instance.

**Comment by PhilippImhof** [comment]
@shinuito Thanks for the hint with STACK's special placeholder. Maybe I can find a way to implement something similar in the future.
